File: negocio/ListaCircularOleadas.java
package negocio;

import modelo.Oleada;

public class ListaCircularOleadas {
    private NodoOleada ultimo;
    private NodoOleada oleadaActual;
    private int tamanio;

    public ListaCircularOleadas() {
        this.ultimo = null;
        this.oleadaActual = null;
        this.tamanio = 0;
    }

    public void registrar(Oleada oleada) {
        NodoOleada nuevo = new NodoOleada(oleada);
        if (ultimo == null) {
            ultimo = nuevo;
            ultimo.setSiguiente(ultimo);
            oleadaActual = ultimo;
        } else {
            nuevo.setSiguiente(ultimo.getSiguiente());
            ultimo.setSiguiente(nuevo);
            ultimo = nuevo;
        }
        tamanio++;
    }

    // FIX: no había forma de saber si un ID de oleada ya existía antes de
    // registrar otra igual (mismo problema que en ListaSecuencialTorres).
    public boolean existeId(int idOleada) {
        if (ultimo == null) return false;
        NodoOleada actual = ultimo.getSiguiente();
        do {
            if (actual.getOleada().getIdOleada() == idOleada) {
                return true;
            }
            actual = actual.getSiguiente();
        } while (actual != ultimo.getSiguiente());
        return false;
    }

    public void mostrar() {
        if (ultimo == null) {
            System.out.println("No hay oleadas registradas.");
            return;
        }
        NodoOleada actual = ultimo.getSiguiente();
        do {
            System.out.println(actual.getOleada());
            actual = actual.getSiguiente();
        } while (actual != ultimo.getSiguiente());
    }

    public Oleada avanzarSiguienteOleada() {
        if (ultimo == null) return null;
        Oleada actual = oleadaActual.getOleada();
        oleadaActual = oleadaActual.getSiguiente();
        return actual;
    }

    public void reiniciarCiclo() {
        if (ultimo != null) {
            oleadaActual = ultimo.getSiguiente();
        }
    }

    public int getTamanio() { return tamanio; }
}

File: Python/negocio/lista_circular_oleadas.py
from negocio.nodo_oleada import NodoOleada


class ListaCircularOleadas:
    """Lista simplemente enlazada circular para las oleadas, igual que
    en la versión Java: al llegar a la última oleada registrada, el
    ciclo vuelve a la primera automáticamente."""

    def __init__(self):
        self.ultimo = None
        self.oleada_actual = None
        self.tamanio = 0

    def registrar(self, oleada):
        nuevo = NodoOleada(oleada)
        if self.ultimo is None:
            self.ultimo = nuevo
            self.ultimo.siguiente = self.ultimo
            self.oleada_actual = self.ultimo
        else:
            nuevo.siguiente = self.ultimo.siguiente
            self.ultimo.siguiente = nuevo
            self.ultimo = nuevo
        self.tamanio += 1

    def existe_id(self, id_oleada):
        if self.ultimo is None:
            return False
        actual = self.ultimo.siguiente
        while True:
            if actual.oleada.id_oleada == id_oleada:
                return True
            actual = actual.siguiente
            if actual is self.ultimo.siguiente:
                break
        return False

    def mostrar(self):
        if self.ultimo is None:
            return ["No hay oleadas registradas."]
        lineas = []
        actual = self.ultimo.siguiente
        while True:
            lineas.append(str(actual.oleada))
            actual = actual.siguiente
            if actual is self.ultimo.siguiente:
                break
        return lineas

    def avanzar_siguiente_oleada(self):
        if self.ultimo is None:
            return None
        actual = self.oleada_actual.oleada
        self.oleada_actual = self.oleada_actual.siguiente
        return actual

    def reiniciar_ciclo(self):
        if self.ultimo is not None:
            self.oleada_actual = self.ultimo.siguiente

    def get_tamanio(self):
        return self.tamanio
